### ISSUE TITLE
How to "consume" messages

### ISSUE THREAD

**Issue by keul** [expressjs/flash#2]
Using the flash package I found an undocumented behavior: flashed messages live forever in the `res.locals.flash` array:

* flash a message
* move to another page
* flash another message

Both messages are flashed. What seems is missing (or not properly documented) is a way to "consume" messages.

**Comment by gluxon** [comment]
I ran into this as well. Fortunately, the source code is easy to understand.

```javascript
req.session.flash = [];
```

**Comment by keul** [comment]
I fixed this as well (while your solution is simpler! Thanks).
Still I think that the original package could provide some native support for this.

**Comment by jonathanong** [comment]
yup. or the way i do is 

```jade
while message = flash.shift()
  .alert.alert-block.alert-error
    p= message
```

i'll add it to the README later, but PR welcomed!

**Comment by keul** [comment]
Thanks!

PS: little off-topic but I'll write here as you will fix the README: there's another typo in the `app.use` usage. You are not providing the `next` argument.

**Comment by alexkavon** [comment]
The flash should be consumed each time a new message is passed or on a new request. That's the point of the flash. Agree/Disagree?

**Comment by dougwilson** [comment]
So a request to load an image should consume the flash?

**Comment by alexkavon** [comment]
Yes. The flash should be empty and the last thing set before sending res.redirect/res.render/etc. So that the template can render the flash. I may be missing something, but it seems to be how a lot of session flash message libraries do it. Instead of shift()-ing template side or constantly writing code to consume messages. What's the point of keeping session messages persistent across multiple requests?

**Comment by keul** [comment]
@alexkavon agree, commonly works that way on other environments

**Comment by alexkavon** [comment]
Should I submit a pull request?

**Comment by dougwilson** [comment]
@alexkavon this repo I defer to @jonathanong for decisions. If you already have the changes, you can submit a PR. If you don't want to work on them without knowing ahead of time, you'll have to wait for @jonathanong :)

**Comment by jonathanong** [comment]
the problem with automatically consuming it is then you'll need an API to figure out, "which requests consume the flash?" 

> What's the point of keeping session messages persistent across multiple requests?

because you only want to delete messages when your user actually reads the message. if you want to only show the last flash message, that's easy, 1-line logic.

**Comment by alexkavon** [comment]
Right so the flash would first be cleared on the next request then you'd populate the flash again with new messages, if there are any. So should I submit a pull?

**Comment by gluxon** [comment]
> Right so the flash would first be cleared on the next request then you'd populate the flash again with new messages, if there are any. So should I submit a pull?

I think you misunderstood @jonathanong's comment. There are two usage cases for flash messages.

1. Show a flash message and consume them on page load.
2. Show a flash message on every page until a user action clears them.

The current implementation caters to usage case 2. I agree with defaulting to this since developers can add 1-line to switch to behavior 1. If we defaulted the other way, developers would have to create their own datastore to manually persist messages if they wanted behavior 2.

I do get your concern with mixing logic on the view layer. If that's too problematic, I would just clear the message session with `req.session.flash = [];`. Otherwise, [connect-flash](https://github.com/jaredhanson/connect-flash) works well. Convincing the developer to adopt a different behavior might not be the best route to go since the current behavior was intended by design and possibly more favorable to some users.

**Comment by alexkavon** [comment]
Perhaps adding a third argument to the req.flash() function or specifying a message type of 'action'  to indicate if user action is needed to clear. There really shouldn't be a need for multiple modules. Also connect-flash hasn't been updated in two years, so I'd like to contribute to helping with an active repo.

**Comment by isimmons** [comment]
Running into this issue now. Here is a single rendering method I have where I have to put req.session.flash = [] twice. I really don't want to have to place this line in every single route or rendering function in my application to make it behave like flash is supposed to behave.

```
exports.renderSignup = function(req, res, next) {
  if (!req.user) {
    req.session.flash = [];
    res.render('signup', {
      title: 'Sign-up Form',
    });

  } else {
    req.session.flash = [];
    return res.redirect('/');
  }
};
```

I'll admit I come from PHP and only have a couple years experience but I have never heard of flash messages that persisted. I thought that was the whole point of calling if flash message instead of persistent message or something.

**Comment by alexkavon** [comment]
@isimmons: The maintainers of this repo want to use connect-flash instead. That's what I've started using.

**Comment by rightaway** [comment]
What's the right place in middleware to put `req.session.flash = []` so that it duplicates the functionality of flash messages in other frameworks?

I tried

```
app.use((req, res, next) => {
  next();
  req.session.flash = [];
});
```

and

```
app.use((req, res, next) => {
  req.session.flash = [];
  next();
});
```

But neither are working.
